### ISSUE TITLE
[Performance] Residual eager backward and batched solve overhead after shared CPU scopes (M5)

### ISSUE THREAD

**Issue by shinaoka** [tensor4all/tenferro-rs#1803]
## What happened?

The M5 CPU refresh after #1802 still exposes substantial eager AD and batched-solve overhead. Track the following three related investigation items separately. These are performance-bug candidates, not established numerical correctness failures.

- [ ] A. Isolate backward cost from unrelated live gradient leaves in the same runtime.
- [ ] B. Measure repeated per-input VJP/source compilation and redundant solve work.
- [ ] C. Explain the prepared-trace batched solve path's approximately 2x forward cost relative to eager.

### A. Backward scans unrelated live leaves (source-confirmed structure; timing attribution pending)

[`backward_from_seed`](https://github.com/tensor4all/tenferro-rs/blob/d8759f4320a337d2399f4a87dfec55af51d2ebf1/crates/tenferro-ad/src/eager.rs#L3982) collects every live gradient slot in the runtime, then calls `vjp_optional` for each candidate. Dependency filtering occurs downstream.

The new benchmark prepares all independent AD fixtures in the same runtime before each batch timer. The 2x2 matmul-backward eager row uses 64 operations per sample, with two gradient leaves per operation. Each backward therefore sees unrelated leaves from the other fixtures. This can introduce quadratic candidate scanning across a batch, even though the intended arithmetic work grows linearly. The source structure is confirmed; a controlled live-leaf sweep has **not** yet measured its contribution. Do not attribute the entire observed eager gap to this mechanism.

### B. Per-input VJP and compilation work remains inside eager backward

The same loop requests VJP separately for each active input. [`semantic_eager_vjp_optional`](https://github.com/tensor4all/tenferro-rs/blob/d8759f4320a337d2399f4a87dfec55af51d2ebf1/crates/tenferro-ad/src/eager.rs#L2711) invokes `compile_ad_source` before looking up the prepared-derivative cache; its comment explicitly says this happens on cache hits to extract bindings.

For forward + sum + backward of 2x2 matmul, 1-thread medians (IQR) are:

| Backend | microseconds per workflow |
|---|---:|
| tenferro eager | 84.847 (1.125) |
| tenferro prepared trace | 19.421 (0.282) |
| PyTorch | 11.199 (0.273) |
| JAX | 3.986 (0.299) |

A prior local LAPACK-interposition investigation on **older** core `28dfc7e383a653e1461d38b73ced3d563b4efdad` observed 4096 `dgetrf` and 4096 `dgetrs` calls for eager forward + backward of 1024 independent 2x2 solves (1024 each in forward, another 3072 each in backward). These call counts have **not** been rerun on `d8759f43` and are only a hypothesis to recheck. Investigate sharing forward LU/X and common adjoint work across requested input gradients, without changing AD semantics.

### C. Prepared-trace batched solve remains substantially slower

Current 4-thread medians (IQR), f64, 2x2 matrices, batch=1024, one RHS:

| Workflow | eager ms | prepared trace ms | PyTorch ms |
|---|---:|---:|---:|
| forward solve | 0.998193 (0.021943) | 1.961927 (0.026938) | 0.059580 (0.003584) |
| forward + sum + backward, both inputs | 4.102209 (0.042166) | 3.640208 (0.019229) | 0.103353 |

The trace forward is approximately 1.97x eager / 32.9x PyTorch. Eager forward+backward is approximately 39.7x PyTorch. Both 1T and 4T show the same pattern.

The older local investigation found trace using `LuFactor + LuSolvePrepared`, retaining factorization but paying pivot/copy/two triangular-solve paths, instead of the eager `dgetrs` path. Recheck this on the current revision and quantify per-matrix allocation/copy/dispatch costs. Do not conflate trace with eager's suspected repeated factorization.

## Expected behavior

A backward request should not pay substantial work proportional to unrelated live graphs in its runtime. Multiple gradients of one output should reuse common analysis/execution where possible. Batched small solves should avoid unnecessary per-matrix copying, allocation and duplicate work. No promise that every small operation benefits from four threads or matches another framework's speed is implied.

## Reproduction and provenance

- Core: `d8759f4320a337d2399f4a87dfec55af51d2ebf1` (clean checkout, #1802 merged).
- Benchmark source: `64970fd383409685da2dbf874a4ad9dcdff4fec0`.
- [Published report, immutable revision](https://github.com/tensor4all/tenferro-benchmark/blob/2718367f9971b509b7523c725ad106c984f268eb/result/mac-cpu/cpu/cpu_ops.md).
- [Benchmark PR and exact collection provenance](https://github.com/tensor4all/tenferro-benchmark/pull/100).
- Apple M5 Max, native macOS 26.5.1 arm64, `system-accelerate`, `TENFERRO_CPU_BACKEND_KIND=blas`, sequential 1/4 threads, idle-host guard enabled.
- PyTorch 2.12.0 / Accelerate; JAX 0.10.1 / XLA CPU. Configured JAX thread settings do not prove actual worker utilization on macOS.

Run in tenferro-benchmark at the source revision above, with `extern/tenferro-rs` at the core revision above:

```bash
BENCHMARK_TARGET_PROFILE=mac-cpu PUBLICATION_GATE_PROFILE=full FFT_BENCH_LENGTHS=1024,16384,1048576 ./scripts/run_all.sh 1 4
```

Inputs, runtime/scope entry, trace preparation and output-retention containers are outside timing. Small cases use calibrated bounded batches and retain outputs until timer stop. `*_backward` rows measure **forward + reduction + backward**, not backward alone. Eager internal AD analysis remains part of that public operation.

Raw CPU-ops CSV and sample JSONL are retained locally under `data/results/mac-cpu/cpu/einsum/20260916_162342` (1T) and `20260916_164245` (4T), not committed. Each thread count has 1016 successful rows. Scope/runtime settings, raw duration normalization and unique keys were checked. Successful timings do not establish numerical parity for every operation; independent reference tests in this benchmark cover batched matmul primal and both input gradients.

## Verification / acceptance criteria

1. Before changing A, hold one tiny backward graph fixed and vary unrelated live gradient leaves (0, 2, 32, 128, 512) in the same runtime. Separately vary batched workflow count. Keep setup/cleanup outside timing and retain every result. Record runtime candidate visits alongside time to distinguish graph work from numerical work.
2. For B, count analysis/compilation, derivative executions and LAPACK factor/solve calls for one vs both requested input gradients, including warmed-cache execution. Verify whether any saved factorization and common reverse work is actually reused.
3. For C, profile eager vs prepared trace on the same solve inputs at sizes 2/4/8/16 and batches 16/64/256/1024. Separate wrapper, allocation/copy, factorization, pivot and triangular-solve work. Keep full results; predeclare comparison/noise gates before any optimization experiment.
4. Validate primal solve residuals and gradients for **both** A and RHS against independent references/finite differences. Include unrelated live graphs and repeated backward calls, preserving gradient accumulation/lifetime contracts.
5. Any fix needs a paired before/after benchmark with identical fixtures, batching and shared scope. Historical vs current report ratios are not a scope-only experiment because fixtures and sampling changed.

## Affected areas and scope

`tenferro-ad` eager runtime/backward/prepared-derivative cache; `tenferro-linalg` trace lowering and CPU solve execution; CPU BLAS/LAPACK batching. Investigate supported existing behavior first. This issue does not authorize new public APIs, dependencies, feature flags, GPU changes or AD semantic changes; proposals requiring those need design acceptance.

Related: #1801 / #1802 (shared scope, now present), #1704 (leaf construction, outside these operation timers), #1628 (older broad Mac report). The earlier several-fold 4T slowdown is absent in these shared-scope CPU-ops results; it is not the primary remaining issue here.
